File: advent-of-code/2023/3/main.cpp
#include <iostream>
#include <vector>
#include <string>
#include <iterator>
#include <algorithm>
#include <fstream>
#include <sstream>
#include <unordered_set>

/**
 * AOC 2023 Day 3
 * Hans Elliott
*/

std::vector<std::string> LINES;
size_t N_LINES = 0;

std::vector<std::pair<int, int>> offsets = {
    {-1, 1}, {-1, 0}, {-1, -1},
    {0, -1}, {0, 1},
    {1, -1}, {1, 0}, {1, 1}
};

std::unordered_set<int> adjacent_nums(int row, int col) {
    std::unordered_set<int> part_nums;
    size_t r, c, numstart, numend;
    unsigned char ch;

    for (const auto &it : offsets) {
        r = row + it.first;
        c = col + it.second;
        if (r < 0 || r >= N_LINES)
            continue;
        if (c < 0 || c >= LINES[r].size())
            continue;
        ch = LINES[r][c];
        if (std::isdigit(ch)) {
            numstart = numend = c;
            while (std::isdigit(LINES[r][numstart - 1]))
                numstart--;
            while (std::isdigit(LINES[r][numend + 1]))
                numend++;
            part_nums.insert(
                std::atoi(LINES[r].substr(numstart, numend + 1).c_str())
            );
        }
    }

    return part_nums;
}


int part1() {
    int partnum_sum = 0;
    unsigned char ch;
    std::unordered_set<int> adjacent;

    for (size_t i = 0; i < N_LINES; i++) {
        for (size_t j = 0; j < LINES[i].size(); j++) {
            ch = LINES[i][j];
            // A symbol is any char that's not a number or a period.
            // Any number adjacent to a symbol is a part number.
            // We need the sum of all part numbers.
            if (!std::isdigit(ch) && ch != '.') {
                adjacent = adjacent_nums(i, j);
                for (const int &num : adjacent) {
                    partnum_sum += num;
                }
            }
        }
    }
    return partnum_sum;
}


int part2() {
    int num_prod, gearratio_sum = 0;
    unsigned char ch;
    std::unordered_set<int> adjacent;

    for (size_t i = 0; i < N_LINES; i++) {
        for (size_t j = 0; j < LINES[i].size(); j++) {

            ch = LINES[i][j];
            // An asterik is a gear if it is adjacent to exactly 2 numbers.
            // The gear's "gear ratio" is the product of those 2 numbers.
            // We need the sum of all the gear ratios.
            if (ch == '*') {
                adjacent = adjacent_nums(i, j);
                if (adjacent.size() == 2) {
                    num_prod = 1;
                    for (const int &num : adjacent) {
                        num_prod *= num;
                    }
                    gearratio_sum += num_prod;
                }
            }
        }
    }
    return gearratio_sum;
}


int main(int argc, char* argv[]) {
    if (argc < 2) {
        std::cerr << "Provide filepath\n";
        return 1;
    }
    std::ifstream infile(argv[1]);
    if (!infile) {
        std::cerr << argv[1] << ": file not found\n";
        return 1;
    }

    /**
     * Read all lines of file into vector so that they can be
     * accessed out of order.
    */
    std::string line;
    while (std::getline(infile, line)) {
        line = "." + line + ".";
        std::istringstream iss(line);
        LINES.emplace_back(std::istream_iterator<char>(iss),
                           std::istream_iterator<char>());
    }
    N_LINES = LINES.size();

    std::cout << "part 1 answer = " << part1() << std::endl;
    std::cout << "part 2 answer = " << part2() << std::endl;

    return 0;
}

File: advent-of-code/2023/3/main.py
#!/usr/bin/python3.11


with open("input.txt", "r") as f:
    # add outer dots so that numbers on the edges get counted
    LINES = [f".{l}." for l in f.read().split("\n")]

N_LINES = len(LINES)


def adjacent_nums(row, col):
    part_nums = set()
    offsets = [(-1,-1), (-1,0), (-1,1),
               (0,-1), (0,1),
               (1,-1), (1,0), (1,1)]
    for y,x in offsets:
        r = row + y
        c = col + x
        if r < 0 or r >= N_LINES:
            continue
        if c < 0 or c >= len(LINES[r]):
            continue
        char = LINES[r][c]
        if char.isdigit():
            numstart = numend = c
            while LINES[r][numstart - 1].isdigit():
                numstart -= 1
            while LINES[r][numend + 1].isdigit():
                numend += 1
            part_nums.add(int(LINES[r][numstart : numend + 1]))
    return part_nums 
    

def part1():
    parts = []
    part_num_sum = 0
    for i, line in enumerate(LINES):
        for j, char in enumerate(line):
            if not char.isdigit() and char != ".":
                part_num_sum += sum(adjacent_nums(i, j))
    # sum of all part numbers (i.e., those adjacent to a symbol)
    return part_num_sum

def part2():
    # gears are "*" symbols adjacent to exactly 2 numbers
    # gear ratio = adjacent num 1 * adjacent num 2
    # we need sum of all gear ratios
    gear_ratio_sum = 0
    for i, line in enumerate(LINES):
        for j, char in enumerate(line):
            if char == "*":
                nums = adjacent_nums(i, j)
                if len(nums) == 2:
                    gear_ratio_sum += nums.pop() * nums.pop()

    return gear_ratio_sum


print(part1())
print(part2())
